### PR TITLE
Unload resource manager

### PR DESCRIPTION
**Pull request by skumbananaaa**
Closes: 
- Make it possible to unload Resources from ResourceManager

### CODE REVIEW

File: LambdaEngine/Source/Resources/ResourceManager.cpp [hunk 1]
```diff
@@ -368,22 +399,83 @@ namespace LambdaEngine
...
+		TextureView* pAmbientOcclusionMapView;
+		TextureView* pMetallicMapView;
+		TextureView* pRoughnessMapView;
 
-		TextureView* pAlbedoMapView				= albedoMap				!= GUID_NONE ? s_TextureViews[albedoMap]			: s_TextureViews[GUID_TEXTURE_DEFAULT_COLOR_MAP];
-		TextureView* pNormalMapView				= normalMap				!= GUID_NONE ? s_TextureViews[normalMap]			: s_TextureViews[GUID_TEXTURE_DEFAULT_NORMAL_MAP];
-		TextureView* pAmbientOcclusionMapView	= ambientOcclusionMap	!= GUID_NONE ? s_TextureViews[ambientOcclusionMap]	: s_TextureViews[GUID_TEXTURE_DEFAULT_COLOR_MAP];
-		TextureView* pMetallicMapView			= metallicMap			!= GUID_NONE ? s_TextureViews[metallicMap]			: s_TextureViews[GUID_TEXTURE_DEFAULT_COLOR_MAP];
-		TextureView* pRoughnessMapView			= roughnessMap			!= GUID_NONE ? s_TextureViews[roughnessMap]			: s_TextureViews[GUID_TEXTURE_DEFAULT_COLOR_MAP];
+		if (albedoMap != GUID_NONE)
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceManager.cpp]: You can make a function or macro out of these if statements. The function/macro would probably accept a GUID and a default/fallback GUID.

File: LambdaEngine/Source/Resources/ResourceManager.cpp [hunk 2]
```diff
@@ -778,7 +886,213 @@ namespace LambdaEngine
...
+
+					//Clean Mesh Name -> Animation GUID
+					s_MeshNamesToAnimationGUIDs.erase(animationsIt);
+				}
+
+				//Clean Mesh GUID -> Name
+				s_MeshGUIDsToNames.erase(meshGUIDToNameIt);
+			}
+
+			return true;
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceManager.cpp]: Move `return true;` up to the end of `if (animationsIt != s_MeshNamesToAnimationGUIDs.end()) { ... }` to only return true if all mappings were found and deleted.

File: LambdaEngine/Source/Resources/ResourceManager.cpp [hunk 3]
```diff
@@ -778,7 +886,213 @@ namespace LambdaEngine
...
+					{
+						UnloadAnimation(animationGUID);
+					}
+
+					//Clean Mesh Name -> Animation GUID
+					s_MeshNamesToAnimationGUIDs.erase(animationsIt);
+				}
+
+				//Clean Mesh GUID -> Name
+				s_MeshGUIDsToNames.erase(meshGUIDToNameIt);
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceManager.cpp]: Move this to right after `auto animationsIt = ...` (line 908) to erase the mapping right after finishing using it. Makes it easier to understand the code.

File: LambdaEngine/Source/Resources/ResourceManager.cpp [hunk 4]
```diff
@@ -778,7 +886,213 @@ namespace LambdaEngine
...
+			}
+
+			return true;
+		}
+
+		LOG_ERROR("[ResourceManager]: Failed to unload Mesh: %d", guid);
+		return false;
+	}
+
+	bool ResourceManager::UnloadMaterial(GUID_Lambda guid)
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceManager.cpp:919]: This function and all unload functions below could return true even if eg. the GUID -> name mapping is broken. To fix this, you could replace the nested scopes with:

`
if (myIt == mapping.end())
{
        return false;
}
`

**Review by TheoBerlin — CHANGES_REQUESTED** [review]
Looks like it should work. Got some comments on readability and error returning. I'm also curious as to where the unload functions should be called from. The destructors of resource classes/structs like `Mesh`?

File: LambdaEngine/Source/Resources/ResourceManager.cpp (hunk 1 above)
Comment (skumbananaaa) [reply LambdaEngine/Source/Resources/ResourceManager.cpp]: Good idea

**Review by TheoBerlin — APPROVED** [review]

**Pull request merged**